proposals/0404-pack-iteration.md

```diff
@@ -0,0 +1,168 @@
+
+```swift
+func == <each Element: Equatable>(lhs: (repeat each Element), rhs: (repeat each Element)) -> Bool {
+
+  for (left, right) in repeat (each lhs, each rhs) {
+    guard left == right else { return false }
+  }
+  return true
+}
+```
+
+The above code iterates pairwise over two tuples `lhs` and `rhs` using a `for-in` loop syntax. At each iteration, the pack elements of `lhs` and `rhs` are bound to the local variables `left` and `right`, respectively. Because `lhs` and `rhs` have the same type, so do `left` and `right`, and the `Equatable` requirement allows comparing `left` and `right` with `==`.
+
+## Detailed Design
+
+In addition to expressions that conform to `Sequence`, the source of a `for-in` loop can be a pack expansion expression. 
+
+```swift
+func iterate<each Element>(over element: repeat each Element) {
+  for element in repeat each element {
+  
+  }
+}
+```
+
+On the *i*th iteration, the type of `element` is the *i*th type parameter in the `Element` type parameter pack, and the value of `element` is the *i*th value parameter in the shadowed `element` value parameter pack. Conceptually, the type of `element` is the pattern type of `repeat each element` with each captured type parameter pack replaced with an implicit scalar type parameter with matching requirements. In this case, the pattern type is `each Element`, so the type of `element` is a scalar type parameter with no requirements. Let’s call the scalar type parameter `Element'`. For example, if `iterate` is called with `each Element` bound to the type pack `{Int, String, Bool}` the body of the `for-in` loop will first substitute `Element'` for `Int`, then `String`, then `Bool`.
+
+If the type parameter packs captured by the pack expansion pattern contain requirements, the scalar type parameter in the loop body will have the same requirements:
+
+```swift
+struct Generic<T> {}
+
+protocol P {}
+
+func iterate<each Element: P>() {
+  for x in repeat Generic<each Element>() {
+    // the type of 'x' is <Element': P> Generic<Element'>
+  }
+}
+```
+
+In the above code, the pattern type of the pack expansion is `Generic<each Element>` where `each Element: P`, so the type of `x` is a scalar type parameter `<Element'> where Element': P`.
+
+Like regular `for-in` loops, `for-in` loops over pack expansions can pattern match over each element of the value pack:
+
+```swift
+enum E<T> {
+  case one(T)
+  case two
+}
+
+func iterate<each Element>(over element: repeat E<each Element>) {
+  for case .one(let value) in repeat each element {
+    // 'value' has type <Element'> Element'
+  }
+}
+```
+
```

Comment:
````suggestion
Unlike iterating over sequences, the pattern expression in the source of a `for-in repeat` loop is evaluated once at each iteration, instead of `n` times eagerly where `n` is the length of the packs captured by the pattern. If `p_i` is the pattern expression at the `i`th iteration and control flow exits the loop at iteration `i`, then `p_j` is not evaluated for `i < j < n`. For example:

```swift
func printAndReturn<Value>(_ value: Value) -> Value {
  print("Evaluated pack element value \(value)")
  return value
}

func iterate<each T>(_ t: repeat each T) {
  var i = 0
  for value in repeat printAndReturn(each t) {
    print("Evaluating loop iteration \(i)")
    if i == 1 { 
      break 
    } else {
      i += 1
    }
  }
  
  print("Done iterating")
}

iterate(1, "hello", true)
```

The above code has the following output

```
Evaluated pack element value 1
Evaluating loop iteration 0
Evaluated pack element value "hello"
Evaluating loop iteration 1
Done iterating
```

````